streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -982,17 +1000,34 @@ private synchronized void setStateWhenNotInPendingShutdown(final State newState)
         setState(newState);
     }
 
-    private synchronized void setState(final State newState) {
-        final State oldState = state;
-        if (!state.isValidTransition(newState)) {
-            log.warn("{} Unexpected state transition from {} to {}.", logPrefix, oldState, newState);
-        } else {
-            log.info("{} State transition from {} to {}.", logPrefix, oldState, newState);
-        }
+    /**
+     * Sets the state
+     * @param newState New state
+     */
+    void setState(final State newState) {
+        synchronized (stateLock) {
+            final State oldState = state;
+
+            // there are cases when we shouldn't check if a transition is valid, e.g.,
+            // when, for testing, a thread is closed multiple times. We could either
+            // check here and immediately return for those cases, or add them to the transition
+            // diagram (but then the diagram would be confusing and have transitions like
+            // PENDING_SHUTDOWN->PENDING_SHUTDOWN).
+            if (newState != State.DEAD && (state == State.PENDING_SHUTDOWN || state == State.DEAD)) {
```

Comment:
Okay. It seems at least one transition we want to skip is that 1) `thread.close()` is called that set its state to `PENDING_SHUTDOWN`, 2) the thread is still in its loop, and in the middle of a rebalance, and likely to try to transit to `ASSIGNING_PARTITIONS` in onPartitionAssigned callback; in this case we do not want to do this transition as the thread is already being notified to shutdown.

I think a better general solution (not necessarily to be piggy-backed in this PR) is to let `setState` return a boolean:

```
// return true: transition is ok and done
// return false: transition is not allowed but not considered fatal error
// throw exception: transition is illegal, and show never be called
boolean setState(final State newState);
```

And in all its callers we do sth. like:

```
if (setState(...)) {
   // proceed normally
} else {
   // handling logic if the transition is not allowed: for example, skip the rest of the callback and proceed.
}
```

cc @dguy